File: brisanet/brisanet_scraper.js
(function() {
    const rawData = JSON.parse(_pageData);
    let results = [];

    function search(obj) {
        if (Array.isArray(obj)) {
            // Nova Assinatura: Verifica se o obj[0] é um ID hexadecimal e obj[1] tem as coordenadas
            if (obj.length >= 6 && typeof obj[0] === 'string' && /^[0-9A-F]{10,}$/i.test(obj[0]) && Array.isArray(obj[1]) && Array.isArray(obj[5])) {

                let id = obj[0];
                let lat = "";
                let lng = "";

                try {
                    // Puxa as coordenadas da nova localização descoberta
                    lat = obj[1][0][0][0];
                    lng = obj[1][0][0][1];
                } catch(e) {}

                // Função interna para garimpar as variáveis exatas dentro desse ponto
                function findAttr(node, key) {
                    let result = "";
                    function dig(n) {
                        if (Array.isArray(n)) {
                            // Se encontrar a etiqueta (ex: "Telefone") e o valor do lado
                            if (n[0] === key && Array.isArray(n[1])) {
                                result = n[1][0];
                                return;
                            }
                            n.forEach(dig);
                        }
                    }
                    dig(node);
                    return result;
                }

                let nome = findAttr(obj[5], "nome") || "";
                let endereco = findAttr(obj[5], "Endereço") || "";
                let telefone = findAttr(obj[5], "Telefone") || "";
                let tipoCadastro = findAttr(obj[5], "Tipo do Cadastro") || "";

                // --- INTELIGÊNCIA PARA FATIAR O ENDEREÇO ---
                // 1. Extrai o CEP (Procura pelo padrão 00000-000)
                let cep = (endereco.match(/\d{5}-\d{3}/) || [""])[0];

                // 2. Extrai UF (Procura por " - CE", " / PB", etc)
                let ufMatch = endereco.match(/[-/]\s*([A-Z]{2})\b/);
                let uf = ufMatch ? ufMatch[1] : "";

                // 3. Extrai Cidade (Pega o texto que vem antes da UF e tenta isolar a cidade)
                let cidade = "";
                if (ufMatch && ufMatch.index > 0) {
                    let textBeforeUF = endereco.substring(0, ufMatch.index);
                    let parts = textBeforeUF.split(/[,\\-]/); // Corta onde tem vírgula ou traço
                    cidade = parts[parts.length - 1].trim(); // Pega a última parte (geralmente a cidade)
                }

                // Limpa quebras de linha e aspas do endereço para não quebrar o CSV
                let enderecoLimpo = endereco.replace(/"/g, "'").replace(/\n/g, " ");

                if (lat && lng) {
                    results.push({
                        id, nome, cidade, uf, endereco: enderecoLimpo, cep, telefone, tipoCadastro, lat, lng
                    });
                }
            }

            // Continua cavando os outros nós
            obj.forEach(search);
        }
    }

    search(rawData);

    if (results.length === 0) {
        console.error("Nada encontrado com a nova estrutura.");
    } else {
        let csv = "ID;Nome da Loja;Cidade;UF;Endereço;CEP;Telefone;Tipo de Cadastro;Latitude;Longitude\n";
        
        results.forEach(r => {
            // CORREÇÃO: Transforma os pontos das coordenadas em vírgulas para o padrão Excel BR
            let latBR = r.lat.toString().replace('.', ',');
            let lngBR = r.lng.toString().replace('.', ',');

            csv += `"${r.id}";"${r.nome}";"${r.cidade}";"${r.uf}";"${r.endereco}";"${r.cep}";"${r.telefone}";"${r.tipoCadastro}";"${latBR}";"${lngBR}"\n`;
        });

        copy(csv);
        console.log("--------------------------------------------");
        console.log(`SUCESSO MÁXIMO! Encontrados ${results.length} registros com todos os dados.`);
        console.log("Os dados já estão no seu Clipboard (Ctrl+C automático).");
        console.log("Vá até a célula A1 do Excel e cole (Ctrl+V)! Agora com coordenadas padrão BR.");
        console.log("--------------------------------------------");
    }
})();

File: brisanet/brisanet_scraper.py
import csv
import json
import re
import sys
from datetime import datetime
from pathlib import Path

import requests


MAP_ID = "1rmfAe5YTRjsPeN9tnkZJzsj3fNGC7QI"
MAP_URL = f"https://www.google.com/maps/d/u/0/viewer?mid={MAP_ID}&femb=1"
BASELINE_RECORDS = 337
MINIMUM_EXPECTED_RECORDS = BASELINE_RECORDS // 2
REQUEST_HEADERS = {"User-Agent": "Mozilla/5.0"}
ROOT_DIR = Path(__file__).resolve().parent
EXTRACTION_DATE = datetime.now().strftime("%Y-%m-%d")
OUTPUT_CSV_FILE = ROOT_DIR / "output" / f"brisanet_Capilaridade_{EXTRACTION_DATE}.csv"
OUTPUT_JSON_FILE = ROOT_DIR / "output" / f"brisanet_Capilaridade_{EXTRACTION_DATE}.json"
CSV_HEADERS = [
    "ID",
    "Nome da Loja",
    "Cidade",
    "UF",
    "Endereço",
    "CEP",
    "Telefone",
    "Tipo de Cadastro",
    "Latitude",
    "Longitude",
]


def find_attr(node, key):
    result = ""

    def dig(value):
        nonlocal result
        if isinstance(value, list):
            if len(value) > 1 and value[0] == key and isinstance(value[1], list):
                result = value[1][0] if value[1] else ""
                return
            for child in value:
                dig(child)

    dig(node)
    return result


def search(obj, results):
    if isinstance(obj, list):
        if (
            len(obj) >= 6
            and isinstance(obj[0], str)
            and re.search(r"^[0-9A-F]{10,}$", obj[0], re.IGNORECASE)
            and isinstance(obj[1], list)
            and isinstance(obj[5], list)
        ):
            item_id = obj[0]
            lat = ""
            lng = ""

            try:
                lat = obj[1][0][0][0]
                lng = obj[1][0][0][1]
            except (IndexError, KeyError, TypeError):
                pass

            nome = find_attr(obj[5], "nome") or ""
            endereco = find_attr(obj[5], "Endereço") or ""
            telefone = find_attr(obj[5], "Telefone") or ""
            tipo_cadastro = find_attr(obj[5], "Tipo do Cadastro") or ""

            cep_match = re.search(r"\d{5}-\d{3}", endereco)
            cep = cep_match.group(0) if cep_match else ""

            uf_match = re.search(r"[-/]\s*([A-Z]{2})\b", endereco)
            uf = uf_match.group(1) if uf_match else ""

            cidade = ""
            if uf_match and uf_match.start() > 0:
                text_before_uf = endereco[: uf_match.start()]
                parts = re.split(r"[,\-]", text_before_uf)
                cidade = parts[-1].strip()

            endereco_limpo = endereco.replace('"', "'").replace("\n", " ")

            if lat and lng:
                results.append(
                    {
                        "id": item_id,
                        "nome": nome,
                        "cidade": cidade,
                        "uf": uf,
                        "endereco": endereco_limpo,
                        "cep": cep,
                        "telefone": telefone,
                        "tipoCadastro": tipo_cadastro,
                        "lat": lat,
                        "lng": lng,
                    }
                )

        for child in obj:
            search(child, results)


def acquire_page_data():
    # verify=True: o alvo e um dominio do Google com cert valido; desabilitar
    # TLS aqui so gerava InsecureRequestWarning repetido e sem motivo.
    response = requests.get(
        MAP_URL, headers=REQUEST_HEADERS, timeout=60
    )
    response.raise_for_status()
    match = re.search(
        r"var\s+_pageData\s*=\s*(\"(?:\\.|[^\"\\])*\")\s*;",
        response.text,
    )
    if not match:
        raise RuntimeError("A variável _pageData não foi encontrada no HTML do viewer.")

    encoded_page_data = json.loads(match.group(1))
    return json.loads(encoded_page_data)


def csv_value(value):
    return str(value).replace(".", ",")


def write_outputs(results):
    OUTPUT_CSV_FILE.parent.mkdir(parents=True, exist_ok=True)
    with OUTPUT_CSV_FILE.open("w", encoding="utf-8-sig", newline="") as output_file:
        output_file.write(";".join(CSV_HEADERS) + "\n")
        for item in results:
            values = [
                item["id"],
                item["nome"],
                item["cidade"],
                item["uf"],
                item["endereco"],
                item["cep"],
                item["telefone"],
                item["tipoCadastro"],
                csv_value(item["lat"]),
                csv_value(item["lng"]),
            ]
            output_file.write(";".join(f'"{value}"' for value in values) + "\n")

    payload = {
        "fonte": MAP_URL,
        "total_lojas": len(results),
        "campos": CSV_HEADERS,
        "lojas": results,
    }
    with OUTPUT_JSON_FILE.open("w", encoding="utf-8") as output_file:
        json.dump(payload, output_file, ensure_ascii=False, indent=2)


def main():
    page_data = acquire_page_data()
    results = []
    search(page_data, results)

    if not results:
        raise RuntimeError(
            "Nenhum registro encontrado: a estrutura de _pageData pode ter mudado."
        )
    if len(results) < MINIMUM_EXPECTED_RECORDS:
        print(
            f"AVISO: foram encontrados {len(results)} registros; "
            f"baseline={BASELINE_RECORDS}.",
            file=sys.stderr,
        )

    write_outputs(results)
    print(f"Sucesso: {len(results)} registros salvos em CSV e JSON.")


if __name__ == "__main__":
    try:
        main()
    except Exception as error:
        print(f"ERRO: {error}", file=sys.stderr)
        sys.exit(1)